sw/device/lib/runtime/epmp.h

```diff
@@ -0,0 +1,458 @@
+ * enum names.
+ */
+typedef enum epmp_permissions {
+  // Unlocked permissions (can always be modified in machine mode).
+  // TODO: do we actually want to have unlocked rules? Could maybe remove
+  // these enums since MML will probably be enabled when in S/U-mode anyway.
+  kEpmpPermissionsUnlockedMachineAllUserNone = 0,    /**< LRWX = 0b0000 */
+  kEpmpPermissionsUnlockedMachineAllUserExecute,     /**< LRWX = 0b0001 */
+  kEpmpPermissionsUnlockedMachineAllUserRead,        /**< LRWX = 0b0100 */
+  kEpmpPermissionsUnlockedMachineAllUserReadExecute, /**< LRWX = 0b0101 */
+  kEpmpPermissionsUnlockedMachineAllUserReadWrite,   /**< LRWX = 0b0110 */
+  kEpmpPermissionsUnlockedMachineAllUserAll,         /**< LRWX = 0b0111 */
+
+  // Locked permissions (can only be modified in machine mode when mseccfg.RLB
+  // is set).
+  kEpmpPermissionsLockedMachineNoneUserNone,               /**< LRWX = 0b1000 */
+  kEpmpPermissionsLockedMachineExecuteUserExecute,         /**< LRWX = 0b1001 */
+  kEpmpPermissionsLockedMachineReadUserRead,               /**< LRWX = 0b1100 */
+  kEpmpPermissionsLockedMachineReadExecuteUserReadExecute, /**< LRWX = 0b1101 */
+  kEpmpPermissionsLockedMachineReadWriteUserReadWrite,     /**< LRWX = 0b1110 */
+  kEpmpPermissionsLockedMachineAllUserAll,                 /**< LRWX = 0b1111 */
+} epmp_permissions_t;
+
+/**
+ * EPMP region specification.
+ *
+ * Provides the start and end addresses of a particular region. These addresses
+ * are byte-aligned (i.e. they are like regular pointers rather than encoded
+ * addresses).
+ *
+ * The intention is that this data structure is used to disambiguate the
+ * addresses of regions. It is therefore recommended that `start` and `end` are
+ * always used as labels when declaring a region.
+ */
+typedef struct epmp_region {
+  uintptr_t start;
+  uintptr_t end;
+} epmp_region_t;
+
+/**
+ * EPMP entry index.
+ */
+typedef enum epmp_entry {
+  kEpmpEntry0,
+  kEpmpEntry1,
+  kEpmpEntry2,
+  kEpmpEntry3,
+  kEpmpEntry4,
+  kEpmpEntry5,
+  kEpmpEntry6,
+  kEpmpEntry7,
+  kEpmpEntry8,
+  kEpmpEntry9,
+  kEpmpEntry10,
+  kEpmpEntry11,
+  kEpmpEntry12,
+  kEpmpEntry13,
+  kEpmpEntry14,
+  kEpmpEntry15,
+} epmp_entry_t;
+
+/**
+ * EPMP entry count.
+ */
+typedef enum epmp_entry_count {
+  kEpmpEntryCount1,
+  kEpmpEntryCount2,
+  kEpmpEntryCount3,
+  kEpmpEntryCount4,
+  kEpmpEntryCount5,
+  kEpmpEntryCount6,
+  kEpmpEntryCount7,
+  kEpmpEntryCount8,
+  kEpmpEntryCount9,
+  kEpmpEntryCount10,
+  kEpmpEntryCount11,
+  kEpmpEntryCount12,
+  kEpmpEntryCount13,
+  kEpmpEntryCount14,
+  kEpmpEntryCount15,
+  kEpmpEntryCount16,
+} epmp_entry_count_t;
+
+/**
+ * EPMP transaction.
+ *
+ * Must be initialized using epmp_transaction_init before use.
+ *
+ * This structure should be considered opaque.
+ */
+typedef struct epmp_transaction {
+  epmp_entry_count_t remaining;
+  epmp_entry_count_t done;
+  epmp_entry_count_t total;  // remaining + done
+  // TODO: add checksum tracker
```

Comment:
This part isn't particularly well thought out. I was thinking it might be useful to verify and update a checksum calculated across the staged register values as part of each configuration step.

Thinking about it now though I wonder if we need to pass the transaction around at all. Since we only allow one transaction at a time anyway it might make sense just to incorporate the transaction state into the staging area.